src/client/pydaos/pydaos_shim.c

```diff
@@ -115,58 +122,109 @@ __shim_handle__err_to_str(PyObject *self, PyObject *args)
  */
 
 static PyObject *
-cont_open(int ret, uuid_t puuid, uuid_t cuuid, int flags)
+cont_open(int ret, char *pool, char *cont, int flags)
 {
-	PyObject	*return_list;
-	daos_handle_t	 coh = {0};
-	daos_handle_t	 poh = {0};
-	int		 rc;
+	PyObject			*return_list;
+	struct open_handle		*hdl = NULL;
+	daos_handle_t			coh = {0};
+	daos_handle_t			poh = {0};
+	daos_prop_t			*prop = NULL;
+	struct daos_prop_entry		*entry;
+	struct daos_prop_co_roots	*roots;
+	int				rc;
 
 	if (ret != DER_SUCCESS) {
 		rc = ret;
 		goto out;
 	}
 
 	/** Connect to pool */
-	rc = daos_pool_connect(puuid, "daos_server", DAOS_PC_RW, &poh,
+	rc = daos_pool_connect(pool, NULL, DAOS_PC_RW, &poh,
 			       NULL, NULL);
 	if (rc)
 		goto out;
 
 	/** Open container */
-	rc = daos_cont_open(poh, cuuid, DAOS_COO_RW, &coh, NULL, NULL);
+	rc = daos_cont_open(poh, cont, DAOS_COO_RW, &coh, NULL, NULL);
 	if (rc)
-		daos_pool_disconnect(poh, NULL);
+		goto out;
+
+	/** Retrieve container properties via cont_query() */
+	prop = daos_prop_alloc(0);
+	if (prop == NULL) {
+		rc = -ENOMEM;
+		goto out;
+	}
+
+	rc = daos_cont_query(coh, NULL, prop, NULL);
+	if (rc)
+		goto out;
+
+	/** Verify that this is a python container */
+	entry = daos_prop_entry_get(prop, DAOS_PROP_CO_LAYOUT_TYPE);
+	if (entry == NULL || entry->dpe_val != DAOS_PROP_CO_LAYOUT_PYTHON) {
+		D_ERROR("Container is not a python container\n");
+		rc = -DER_INVAL;
+		goto out;
+	}
+
+	/** Fetch root object ID */
+	entry = daos_prop_entry_get(prop, DAOS_PROP_CO_ROOTS);
+	if (entry == NULL) {
+		D_ERROR("Invalid entry in properties for root object ID\n");
+		rc = -DER_INVAL;
+		goto out;
+	}
+	roots = (struct daos_prop_co_roots *)entry->dpe_val_ptr;
+	if (roots->cr_oids[0].hi == 0 && roots->cr_oids[0].lo == 0) {
+		D_ERROR("Invalid root object ID in properties\n");
+		rc = -DER_INVAL;
+		goto out;
+	}
+
+	/** Track container open handle and associated attributes */
+	D_ALLOC_PTR(hdl);
+	if (hdl == NULL) {
+		D_ERROR("failed to allocate internal handle to open container\n");
+		rc = -DER_NOMEM;
+		goto out;
+	}
+	hdl->poh	= poh;
+	hdl->coh	= coh;
+	hdl->root	= roots->cr_oids[0];
+	/** Use flatkv option for root kv */
+	hdl->root.hi	|= (uint64_t)DAOS_OF_KV_FLAT << OID_FMT_FEAT_SHIFT;
```

Comment:
this should already be set when you store the oid in roots?

Reply:
I did not set DAOS_OF_KV_FLAT when calling daos_obj_generate_oid_by_rf(). Shall I?

Reply:
you did?
https://github.com/daos-stack/daos/pull/6396/files#diff-0a19604e5ee8ebcf3fc7581699cc1bc08168d690670f7f4951601d0d62529a0aR473
maybe i missed another occurency

Reply:
ah wait.. im confused here.. shouldn't the roots oid be set in the python layer?
im not sure why you are setting it in the generic props create.

Reply:
This code is for the KV oid allocation and not the root oids.
Correct, i generated it in cont create if no ROOTS property is specified by the caller. I was thinking that this is something that might be helpful. Worst case, the middleware does not use it.
If you think that this is not appropriate, i can just move it to daos cont create when the container type is PYTHON. Just let me know.

Reply:
I am fine with the change you made to daos_cont_create for just storing a generic root oid if not specified.
But for this case, i think the roots oid stored should include the feat flag, so yea we could update the daos_cont_create to add it for python type container. or we could just make this change in the daos_hdlr.c and have a case for python type container, but that handles the case only when the user creates the container with the tool.

Reply:
Sounds good. I will fix it in a subsequent patch.